### PR TITLE
VEGA-538: Switch to permissions-based authorization

### PR DESCRIPTION
**Pull request by gregtyler**
Use permissions to determine what handlers a user should have access to.

Also tidy up our permission checking: use GetMyPermissions to retrieve, then `.HasPermission` on the returned PermissionSet with constants for verbs.

### CODE REVIEW

File: internal/server/edit_my_details.go [hunk 1]
```diff
@@ -9,7 +9,7 @@ import (
 type EditMyDetailsClient interface {
 	MyDetails(sirius.Context) (sirius.MyDetails, error)
 	EditMyDetails(sirius.Context, int, string) error
-	HasPermission(sirius.Context, string, string) (bool, error)
+	GetMyPermissions(sirius.Context) (sirius.PermissionSet, error)
```
Comment (hawx) [inline internal/server/edit_my_details.go:12]: The usual naming for getters is to not use `Get`, so `MyPermissions` here https://golang.org/doc/effective_go.html#Getters

File: internal/server/require_permissions.go [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+type RequirePermissionClient interface {
+	GetMyPermissions(sirius.Context) (sirius.PermissionSet, error)
+}
+
+type PermissionRequest struct {
+	group  string
+	method string
+}
+
+func requirePermissions(client RequirePermissionClient, permissions ...PermissionRequest) func(Handler) Handler {
```
Comment (hawx) [inline internal/server/require_permissions.go:18]: you only ever pass in 1 `PermissionRequest`

File: internal/sirius/get_my_permissions_test.go [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+		})
+	}
+}
+
+func TestPermissionsIgnoredPact(t *testing.T) {
+	// We need this test to produce a specific array of permissions in the
+	// response so that Cypress tests will pass. Since Pact won't let us return
+	// multiple array entries from `dsl.EachLike` we have to write a separate
+	// test with the specific output. We call the consumer "z-append" to ensure
+	// the Pact is loaded last and overwrites the actual test of /api/permission
```
Comment (hawx) [inline internal/sirius/get_my_permissions_test.go:113]: Interesting way of doing this, I'm not sure how I feel about using a test to overwrite mock data. I guess the pro is that you know the code handles this case our side... 🤔 

Maybe I should have done the teams one like this

**Review by hawx — COMMENTED** [review]
One of the things that wasn't so great about the way we did permissions before was that the definitions were basically untested, this has the same drawback. Plus if we want permissions in the handler we have to call the endpoint again.

I was wondering if we extend `Handler` to be `type Handler func(perm PermissionSet, w http.ResponseWriter, r *http.Request) error`, we could then push the checks down to the handler level and test as normal. We'd then only need to retrieve the permissions once, in `errorHandler`.

Thoughts?

File: internal/sirius/get_my_permissions_test.go (hunk 3 above)
Comment (gregtyler) [reply internal/sirius/get_my_permissions_test.go:113]: I tried a few approaches and most of them just aren't supported by Pact. I suspect the proper answer is to not rely on _just_ Pact for all our integration tests, but that model works for us now and I don't want to mess with it.

File: internal/server/add_team_member_test.go [hunk 4]
```diff
@@ -58,6 +58,10 @@ func (c *mockAddTeamMemberClient) SearchUsers(ctx sirius.Context, search string)
 	return c.searchUsers.data, c.searchUsers.err
 }
 
+func (c *mockAddTeamMemberClient) requiredPermissions() sirius.PermissionSet {
```
Comment (hawx) [inline internal/server/add_team_member_test.go:61]: could have instead made this a top-level var, but this also works

**Review by hawx — APPROVED** [review]

**Pull request merged**